pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageListenerExecutor.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.io.Serializable;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Interface for providing an executor service to execute message listeners.
+ */
+public interface MessageListenerExecutor extends Serializable {
+
+    ExecutorService getExecutor();
+
+    /**
+     * Execute the runnable with the given shard key. The runnable will be executed by same thread of the same shard key.
+     *
+     * @param shardKey the shard key
+     * @param runnable the runnable to execute
+     */
+    void execute(byte[] shardKey, Runnable runnable);
```

Comment:
> Instead of adding `shardKey`, simply add the `Message` parameter. This will allow the implementation to do whatever ordering is needed for the use case. For example, a priority queue could be used based on some value in the message headers etc.

Thanks, got it !

Reply:
> Instead of adding `shardKey`, simply add the `Message` parameter. This will allow the implementation to do whatever ordering is needed for the use case. For example, a priority queue could be used based on some value in the message headers etc.

By the way, is it necessary to add `SubscriptionType` parameter? For example, the Key_Shared SubscriptionType need to select a thread by order key.

```suggestion
    void execute(SubscriptionType subscriptionType, Message<?> message,  Runnable runnable);
```

Reply:
I don't think that it's necessary. The application will have that context and could pass that to the implementation of the interface by other means. It doesn't change for each call.

Reply:
> I don't think that it's necessary. The application will have that context and could pass that to the implementation of the interface by other means. It doesn't change for each call.

Okay!